### PR TITLE
Craete noctx

### PR DESCRIPTION
**Pull request by sonatard**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,90 @@
...
+	req4, _ := http.NewRequest(http.MethodPost, url, nil) // want `Must not use http.NewRequest. Use http.NewRequestWithContext or http.NewRequest with \(\*Request\)\.WithContext`
+	req4 = f2(req4, ctx)
+	cli.Do(req4)
+}
+```
+
+## Reference
+- [net/http - NewRequest](https://golang.org/pkg/net/http/#NewRequest)
+- [net/http - NewRequestWithContext](https://golang.org/pkg/net/http/#NewRequestWithContext)
+- [net/http - Request.WithCOntext](https://golang.org/pkg/net/http/#Request.WithContext)
```
Comment (tenntenn) [inline README.md]: withContext

File: reqwithoutctx/ssa.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	return reqExts
+}
+
+func extractByCall(call *ssa.Call) []*ssa.Extract {
+	var exts []*ssa.Extract
+
+	fType := call.String()
+
+	// skip net/http.Request method call
+	if strings.Contains(fType, "net/http.Request).") {
```
Comment (tenntenn) [inline reqwithoutctx/ssa.go]: use types.Identical

File: reqwithoutctx/ssa.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+		if ext, ok := arg.(*ssa.Extract); ok && strings.Contains(ext.Type().String(), "net/http.Request") {
+			exts = append(exts, ext)
+		}
+	}
+
+	return exts
+}
+
+func extractByUnOp(op *ssa.UnOp) *ssa.Extract {
+	if ext, ok := op.X.(*ssa.Extract); ok && strings.Contains(ext.Type().String(), "net/http.Request") {
```
Comment (tenntenn) [inline reqwithoutctx/ssa.go]: use types.Identical

File: reqwithoutctx/ssa.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+	return nil
+}
+
+func extractByReturn(ret *ssa.Return) []*ssa.Extract {
+	rets := ret.Results
+	exts := make([]*ssa.Extract, 0, len(rets))
+
+	for _, ret := range rets {
+		if ext, ok := ret.(*ssa.Extract); ok && strings.Contains(ext.Type().String(), "net/http.Request") {
```
Comment (tenntenn) [inline reqwithoutctx/ssa.go]: use types.Identical

File: testdata/src/a/a.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	}()
+
+	func() (*http.Request, *http.Request) {
+		req14, _ := http.NewRequest(http.MethodPost, url, nil) // want `should rewrite http.NewRequest to http.NewRequestWithContext or http.NewRequest and \(\*Request\).WithContext`
+		req15, _ := http.NewRequest(http.MethodPost, url, nil) // OK
+		req15 = req15.WithContext(ctx)
+
+		return req14, req15
+	}()
+}
```
Comment (tenntenn) [inline testdata/src/a/a.go:132]: add 

```go
_, _ = func() *http.Request {return req}(), nil //OK
_, _ = func() *http.Request {return req2}(), nil //OK
```

File: ngfunc/report.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package ngfunc
+
+import (
+	"fmt"
+	"go/token"
+	"go/types"
+
+	"golang.org/x/tools/go/analysis"
+
```
Comment (tenntenn) [inline ngfunc/report.go]: remove the blank line

File: reqwithoutctx/ssa.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	reqs := make(map[*ssa.Call]*ssa.Extract)
+
+	newRequestType := analysisutil.TypeOf(pass, "net/http", "NewRequest")
+
+	srcFuncs := pass.ResultOf[buildssa.Analyzer].(*buildssa.SSA).SrcFuncs
+	for _, f := range srcFuncs {
+		for _, b := range f.Blocks {
+			for _, instr := range b.Instrs {
+				if ext, ok := instr.(*ssa.Extract); ok {
+					if ext.Type().String() == "*net/http.Request" {
```
Comment (tenntenn) [inline reqwithoutctx/ssa.go]: use types.Identical

File: ngfunc/main.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+import (
+	"go/types"
+
+	"github.com/gostaticanalysis/analysisutil"
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/buildssa"
+)
+
+func Run(pass *analysis.Pass) (interface{}, error) {
+	ngFuncNames := []string{
```
Comment (tenntenn) [inline ngfunc/main.go:12]: It can skip when the package doesn't import net/http.

File: reqwithoutctx/ssa.go [hunk 9]
```diff
@@ -88,31 +86,58 @@ func extractByCall(call *ssa.Call) []*ssa.Extract {
...
 }
+
+func (a *analyzer) requestsByNewRequest() map[*ssa.Call]*ssa.Extract {
+	reqs := make(map[*ssa.Call]*ssa.Extract)
+
+	srcFuncs := a.pass.ResultOf[buildssa.Analyzer].(*buildssa.SSA).SrcFuncs
+	for _, f := range srcFuncs {
+		for _, b := range f.Blocks {
+			for _, instr := range b.Instrs {
+				if ext, ok := instr.(*ssa.Extract); ok {
```
Comment (tenntenn) [inline reqwithoutctx/ssa.go]: Change conditions to reduce nests.
```
ext, ok := instr.(*ssa.Extract)
if !ok {
    continue
}

if !types.Identical(ext.Type(), a.requestType) {
    continue
}

...
```

File: reqwithoutctx/ssa.go [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+type Analyzer struct {
+	netHTTPImportFuncs []*ssa.Function
+	newRequestType     types.Type
+	requestType        types.Type
+}
+
+func NewAnalyzer(pass *analysis.Pass) *Analyzer {
+	newRequestType := analysisutil.TypeOf(pass, "net/http", "NewRequest")
+	requestType := analysisutil.TypeOf(pass, "net/http", "*Request")
```
Comment (tenntenn) [inline reqwithoutctx/ssa.go:21]: You can skip analyzing if `newRequestType` or `requestType` are `nil`.

File: reqwithoutctx/ssa.go (hunk 2 above)
Comment (sonatard) [reply reqwithoutctx/ssa.go]: 直しました！

**Pull request merged**